### PR TITLE
Accept LA icons for the toolbar

### LINKED ISSUE #25164: [Bug]: LA image mode not working anymore for custom toolbar buttons
### Bug summary

Prior to `matplotlib 3.6`, I was able to use an icon for a custom tool in the toolbar of a figure which had a `LA` mode (see attached). Now I can't seem to be able to load it anymore as a custom icon.
icon

### Code for reproduction

```python
import os

import matplotlib
matplotlib.rcParams["toolbar"] = "toolmanager"
import matplotlib.pyplot as plt
from matplotlib.backend_tools import ToolToggleBase


class CustomTool(ToolToggleBase):
    image = os.path.join(os.path.dirname(__file__), "icon.png")


if __name__ == "__main__":
    fig = plt.figure()
    toolmanager = fig.canvas.manager.toolmanager
    toolbar = fig.canvas.manager.toolbar

    toolmanager.add_tool("custom", CustomTool)
    toolbar.add_tool("custom", "group")
```

### Actual outcome

```
/home/[...]/mwe.py:4: UserWarning: Treat the new Tool classes introduced in v1.5 as experimental for now; the API and rcParam may change in future versions.
  matplotlib.rcParams["toolbar"] = "toolmanager"
Traceback (most recent call last):
  File "/home/[...]/mwe.py", line 19, in <module>
    toolbar.add_tool("custom", "group")
  File "/home/[...]/lib/python3.10/site-packages/matplotlib/backend_bases.py", line 3377, in add_tool
    self.add_toolitem(tool.name, group, position,
  File "/home/[...]/lib/python3.10/site-packages/matplotlib/backends/_backend_tk.py", line 957, in add_toolitem
    button = NavigationToolbar2Tk._Button(frame, name, image_file, toggle,
  File "/home/[...]/lib/python3.10/site-packages/matplotlib/backends/_backend_tk.py", line 792, in _Button
    NavigationToolbar2Tk._set_image_for_button(self, b)
  File "/home/[...]/lib/python3.10/site-packages/matplotlib/backends/_backend_tk.py", line 734, in _set_image_for_button
    im_alt = _recolor_icon(im, foreground)
  File "/home/[...]/lib/python3.10/site-packages/matplotlib/backends/_backend_tk.py", line 722, in _recolor_icon
    image_data[black_mask, :3] = color
ValueError: shape mismatch: value array of shape (3,) could not be broadcast to indexing result of shape (0,2)
```

### Expected outcome

No errors should be raised and the icon should be displayed in the toolbar.

### Additional information

Downgrading to `matplotlib v 3.5.3` solves the issue. Don't know why this happens, but icons that have a `RGBA` mode (like the default icons) don't seem to have this problem. The attached icon has a `LA` mode.

### Operating system

Ubuntu 22.04

### Matplotlib Version

3.6.3

### Matplotlib Backend

TkAgg

### Python version

3.10.6

### Jupyter version

N/A

### Installation

pip

### PR DESCRIPTION
**Pull request by fgoudreault**
## PR Summary
Fixes #25164 by converting a PIL icon image to RGBA when loading it. Also created unittest to check the bug in the future.

## PR Checklist

**Documentation and Tests**
- [X] Has pytest style unit tests (and `pytest` passes)
- [ ] [ N/A ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] [ N/A ] New plotting related features are documented with examples.

**Release Notes**
- [ ] [ N/A ] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [ ] [ N/A ] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [ ] [ N/A ] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

### CODE REVIEW

File: lib/matplotlib/tests/test_backend_tk.py
Comment (jklymak) [inline lib/matplotlib/tests/test_backend_tk.py:193]: These imports should be at the top?

File: lib/matplotlib/tests/test_backend_tk.py
Comment (jklymak) [inline lib/matplotlib/tests/test_backend_tk.py:213]: stray print statement

**Review by jklymak — COMMENTED** [review]
Looks good.  not sure, but should this be in `test_backend_tools.py`?

File: lib/matplotlib/tests/test_backend_tk.py
Comment (fgoudreault) [reply lib/matplotlib/tests/test_backend_tk.py:213]: No because the `@_isolated_tk_test(success_count=1)` decorator looks for these print statement to mark the test as success by parsing the `stderr`. I could remove it and but `success_count=0`. Would that be preferable?

File: lib/matplotlib/tests/test_backend_tk.py
Comment (fgoudreault) [reply lib/matplotlib/tests/test_backend_tk.py:193]: sure I was just following the same "coding style" as the other tests in this file.

File: lib/matplotlib/tests/test_backend_tk.py
Comment (jklymak) [reply lib/matplotlib/tests/test_backend_tk.py:213]: OK, I'd never seen a test structured that way, but if you are sure it's correct, no problem.  Maybe add a comment?

File: lib/matplotlib/tests/test_backend_tk.py
Comment (jklymak) [reply lib/matplotlib/tests/test_backend_tk.py:193]: Again, not a problem if thats how the rest of the tests are in this file - definitely an idiosyncratic structure!  OTOH, are we sure we don't want this in `test_tools_backend.py`?

File: lib/matplotlib/tests/test_backend_tk.py
Comment (fgoudreault) [reply lib/matplotlib/tests/test_backend_tk.py:213]: I'll get rid of the print

File: lib/matplotlib/tests/test_backend_tk.py
Comment (fgoudreault) [reply lib/matplotlib/tests/test_backend_tk.py:193]: sorry I replied in the main thread for this question

**Comment by jklymak** [comment]
> > Looks good. not sure, but should this be in `test_backend_tools.py`?
> 
> I put the test in `test_backend_tk.py` as this was a bug related only to `TkAgg` AFAIK. The actual bug fix is in `_backends/backend_tk.py`.

Thats probably fine, but is the same bug present in other backends?

**Comment by fgoudreault** [comment]
> > > Looks good. not sure, but should this be in `test_backend_tools.py`?
> > 
> > 
> > I put the test in `test_backend_tk.py` as this was a bug related only to `TkAgg` AFAIK. The actual bug fix is in `_backends/backend_tk.py`.
> 
> Thats probably fine, but is the same bug present in other backends?

As the bug stems from the `backend_tk.py` file It would seem the bug would only affect this backend. Additionally, I just tested with `qtagg` and PyQt6 and there is no bug on the `main` branch.

**Comment by jklymak** [comment]
Thats great - however, does the test apply to the other backends?  If so, perhaps better to put somewhere that all the backends get tested to increase our test coverage?    BTW, I haven't checked all this, so feel free to tell me the test is definitely only useful for Tk and I'm happy to approve - we can always generalize the test later if you are not correct.

**Comment by fgoudreault** [comment]
Indeed, this test could be applied to any backend! You are right that it would be more useful if I can parametrize the test for any backend. Although I am not familiar with the whole set of matplotlib unittests. Do you know if there are tests that parametrize over backends so that I can generalize the one I introduce here? Thanks

**Comment by fgoudreault** [comment]
> Indeed, this test could be applied to any backend! You are right that it would be more useful if I can parametrize the test for any backend. Although I am not familiar with the whole set of matplotlib unittests. Do you know if there are tests that parametrize over backends so that I can generalize the one I introduce here? Thanks

I used the parametrization defined in `test_backends_interactive.py` in order to get environments with different backends and run them in a subprocess helper function. The parametrization now gives ~13 tests for different configuration. I tested with `pyqt6`, `pyside6` and `pycairo` installed in my environment and it seems to work. I couldn't make it work with `pyqt5` and other `pysides` as I had trouble installing these packages.

**Comment by fgoudreault** [comment]
Oh it seems there is a very similar bug in the wx backend seen on appveyor. I could reproduce it locally and fixed it. I'll push the fix.

**Comment by anntzer** [comment]
Sorry for the conflicting informations, but I think this test should rather go into test_backends_interactive.py::_test_interactive_impl.  This is because interactive tests are very, very slow to start up (as they need to be run into their own subprocess) so it is better to try and minimize the number of tests and do more in a single test.  Probably we should better document this at the top of test_backends_interactive.py...

**Comment by fgoudreault** [comment]
> Sorry for the conflicting informations, but I think this test should rather go into test_backends_interactive.py::_test_interactive_impl. This is because interactive tests are very, very slow to start up (as they need to be run into their own subprocess) so it is better to try and minimize the number of tests and do more in a single test. Probably we should better document this at the top of test_backends_interactive.py...

Sure I'll move it there

**Review by jklymak — APPROVED** [review]
I'm not an expert on how the tests are structures (obviously :sheep:), but this looks good to me.  Perhaps @anntzer will have a minute for a second review.

File: lib/matplotlib/tests/test_backends_interactive.py
Comment (tacaswell) [inline lib/matplotlib/tests/test_backends_interactive.py:203]: Just to check, this changes an E result to a F result?

Comment (fgoudreault) [reply lib/matplotlib/tests/test_backends_interactive.py:203]: I don't believe so since when an error is raised, the test results in a failure and do not go through the assert statement defined after. Therefore no `E` in any case.

Comment (tacaswell) [reply lib/matplotlib/tests/test_backends_interactive.py:203]: Then why add this try/except?

**Review by tacaswell — APPROVED** [review]
Modulo sorting out the extra try/except added in the test.

File: lib/matplotlib/tests/test_backends_interactive.py
Comment (fgoudreault) [reply lib/matplotlib/tests/test_backends_interactive.py:203]: This is to print out the cause of the error as printed out in `stderr` otherwise we only see a traceback of the subprocess error which is not very helpful when debugging.
P.S.: the same `try/except` is done in the `test_backend_tk.py` [tests](https://github.com/matplotlib/matplotlib/blob/main/lib/matplotlib/tests/test_backend_tk.py#L45) to extract the error statements.

Comment (tacaswell) [reply lib/matplotlib/tests/test_backends_interactive.py:203]: might be worth pushing that logic into our helper, but not going to hold this PR over that!

**Comment by tacaswell** [comment]
Thank you for your work on this @fgoudreault and congratulations on your first merged Matplotlib PR 🎉 I hope we hear from you again!

**Pull request merged**
